File: Payroll.java
abstract class AbstractPerson {
    private String name;
    private String phoneNumber;

    public AbstractPerson(String name, String phoneNumber) {
        this.name = name;
        this.phoneNumber = phoneNumber;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getPhoneNumber() {
        return phoneNumber;
    }

    public void setPhoneNumber(String phoneNumber) {
        this.phoneNumber = phoneNumber;
    }

    public abstract boolean login(String phoneNumber);
}

class FitnessTrainer extends AbstractPerson {
    private String specialization;

    public FitnessTrainer(String name, String phoneNumber, String specialization) {
        super(name, phoneNumber);
        this.specialization = specialization;
    }

    public String getSpecialization() {
        return specialization;
    }

    public void setSpecialization(String specialization) {
        this.specialization = specialization;
    }

    public void leadTrainingSession(String exercise) {
        // Logic to lead a training session
        System.out.println(getName() + " is leading a training session on " + exercise);
    }

    @Override
    public boolean login(String phoneNumber) {
        // Logic for trainer login
        return true; // Placeholder for actual logic
    }
}

class Client extends AbstractPerson {
    private FitnessCenterProgram program;
    private String currentExercise;

    public Client(String name, String phoneNumber, FitnessCenterProgram program) {
        super(name, phoneNumber);
        this.program = program;
    }

    public FitnessCenterProgram getProgram() {
        return program;
    }

    public void setProgram(FitnessCenterProgram program) {
        this.program = program;
    }

    public String getCurrentExercise() {
        return currentExercise;
    }

    public void setCurrentExercise(String currentExercise) {
        this.currentExercise = currentExercise;
    }

    public void performExercise(String exercise) {
        // Logic to perform exercise
        System.out.println(getName() + " is performing " + exercise);
    }

    public void recordExercise(String exercise) {
        // Logic to record exercise
        System.out.println(getName() + " has recorded exercise: " + exercise);
    }

    public double calculateCaloriesBurned(double duration) {
        // Logic to calculate calories burned
        return duration * 10; // Example calculation
    }

    public void sendProgressToTrainer() {
        // Logic to send progress to trainer
        System.out.println(getName() + " is sending progress to trainer.");
    }

    @Override
    public boolean login(String phoneNumber) {
        // Logic for client login
        return true; // Placeholder for actual logic
    }
}

class FitnessCenter {
    private String location;
    private double membershipFee;

    public FitnessCenter(String location, double membershipFee) {
        this.location = location;
        this.membershipFee = membershipFee;
    }

    public String getLocation() {
        return location;
    }

    public void setLocation(String location) {
        this.location = location;
    }

    public double getMembershipFee() {
        return membershipFee;
    }

    public void setMembershipFee(double membershipFee) {
        this.membershipFee = membershipFee;
    }

    public void requestTrainingSession(FitnessTrainer trainer, String exercise) {
        // Logic to request a training session
        System.out.println("Requesting training session with " + trainer.getName() + " for exercise: " + exercise);
    }
}

class FitnessCenterProgram {
    private String programName;
    private String version;

    public FitnessCenterProgram(String programName, String version) {
        this.programName = programName;
        this.version = version;
    }

    public String getProgramName() {
        return programName;
    }

    public void setProgramName(String programName) {
        this.programName = programName;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    public void recordExercise(String exercise) {
        // Logic to record exercise
        System.out.println("Recording exercise: " + exercise);
    }

    public double calculateCaloriesBurned(double duration) {
        // Logic to calculate calories burned
        return duration * 10; // Example calculation
    }

    public void sendProgressToTrainer() {
        // Logic to send progress to trainer
        System.out.println("Sending progress to trainer.");
    }
}

public class Payroll {
    public static void main(String[] args) {
        // Example usage
        FitnessCenterProgram program = new FitnessCenterProgram("Weight Loss", "1.0");
        Client client = new Client("John Doe", "123-456-7890", program);
        FitnessTrainer trainer = new FitnessTrainer("Jane Smith", "098-765-4321", "Yoga");

        client.performExercise("Push Ups");
        client.recordExercise("Push Ups");
        client.sendProgressToTrainer();
        trainer.leadTrainingSession("Yoga");

        FitnessCenter fitnessCenter = new FitnessCenter("Main Street Gym", 50.0);
        fitnessCenter.requestTrainingSession(trainer, "Yoga");
    }
}

File: payroll.py
from abc import ABC, abstractmethod


class AbstractPerson(ABC):
    def __init__(self, name: str, phone_number: str):
        self._name = name
        self._phone_number = phone_number

    @property
    def name(self):
        return self._name

    @name.setter
    def name(self, value):
        self._name = value

    @property
    def phone_number(self):
        return self._phone_number

    @phone_number.setter
    def phone_number(self, value):
        self._phone_number = value

    @abstractmethod
    def login(self, phone_number: str) -> bool:
        pass


class FitnessTrainer(AbstractPerson):
    def __init__(self, name: str, phone_number: str, specialization: str):
        super().__init__(name, phone_number)
        self.specialization = specialization

    def lead_training_session(self, exercise: str):
        print(f"{self.name} is leading a training session on {exercise}")

    def login(self, phone_number: str) -> bool:
        # Logic for trainer login
        return True  # Placeholder for actual logic


class FitnessCenterProgram:
    def __init__(self, program_name: str, version: str):
        self.program_name = program_name
        self.version = version

    def record_exercise(self, exercise: str):
        print(f"Recording exercise: {exercise}")

    def calculate_calories_burned(self, duration: float) -> float:
        return duration * 10  # Example calculation

    def send_progress_to_trainer(self):
        print("Sending progress to trainer.")


class Client(AbstractPerson):
    def __init__(self, name: str, phone_number: str, program: FitnessCenterProgram):
        super().__init__(name, phone_number)
        self.program = program
        self.current_exercise = None

    def perform_exercise(self, exercise: str):
        print(f"{self.name} is performing {exercise}")

    def record_exercise(self, exercise: str):
        print(f"{self.name} has recorded exercise: {exercise}")

    def calculate_calories_burned(self, duration: float) -> float:
        return duration * 10  # Example calculation

    def send_progress_to_trainer(self):
        print(f"{self.name} is sending progress to trainer.")

    def login(self, phone_number: str) -> bool:
        # Logic for client login
        return True  # Placeholder for actual logic


class FitnessCenter:
    def __init__(self, location: str, membership_fee: float):
        self.location = location
        self.membership_fee = membership_fee

    def request_training_session(self, trainer: FitnessTrainer, exercise: str):
        print(f"Requesting training session with {trainer.name} for exercise: {exercise}")


# Example usage
if __name__ == "__main__":
    program = FitnessCenterProgram("Weight Loss", "1.0")
    client = Client("John Doe", "123-456-7890", program)
    trainer = FitnessTrainer("Jane Smith", "098-765-4321", "Yoga")

    client.perform_exercise("Push Ups")
    client.record_exercise("Push Ups")
    client.send_progress_to_trainer()
    trainer.lead_training_session("Yoga")

    fitness_center = FitnessCenter("Main Street Gym", 50.0)
    fitness_center.request_training_session(trainer, "Yoga")
